### PR TITLE
[3.9] Restrict hybrid disjoint sg directions for Path Queries

### PR DESCRIPTION
**Pull request by mchacki**
### Scope & Purpose

Backport of https://github.com/arangodb/arangodb/pull/16214

This PR hits the enterprise version only.
For Hybrid Disjoint SmartGraphs we decided to restrict switches between Smart and Satellite
for a single AQL Node to at most one switch.
So every Graph node that switches two (or more) times will cause the query to reject.
Note: **This is a breaking change!** there can be queries that do not work anymore.
However original behavior especially for path queries, delivered undefined results.

Please see Enterprise PR for technical details.

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [x] Tests
  - [x] **Regression tests**
  - [ ] C++ **Unit tests**
  - [x] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [x] Backports
  - [x] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*
  - [ ] Backport for 3.7: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/946
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: CHANGELOG [hunk 1]
```diff
@@ -8,6 +8,23 @@ v3.9.2 (XXXX-XX-XX)
 * Make all requests which are needed for shard resync at least medium priority
   to improve getting-in-sync under load.
 
+* Enterprise only: Restricted behaviour of Hybrid Disjoint Smart Graphs. Within a single
+  traversal or path query we now restrict that you can only switch between
+  Smart and Satellite sharding once, all queries where  more than one switch
+  is (in theory) possible will be rejected. e.g:
```
Comment (jsteemann) [inline CHANGELOG:14]:
```suggestion
* Enterprise only: Restricted behavior of Hybrid Disjoint Smart Graphs. Within a 
  single traversal or path query we now restrict that you can only switch between
  Smart and Satellite sharding once, all queries where more than one switch
  is (in theory) possible will be rejected. e.g:
```

File: CHANGELOG [hunk 2]
```diff
@@ -8,6 +8,23 @@ v3.9.2 (XXXX-XX-XX)
...
+    FOR v IN 2 OUTBOUND @start smartToSatEdges, satToSmartEdges
+  ```
+  will be rejected (we can go smart -> sat -> smart, so two switches)
+  ```
+  FOR v1 IN 1 OUTBOUND @start smartToSatEdges
+    FOR v2 IN 1 OUTBOUND v1 satToSmartEdges
+  ```
+  will still be allowed, as each statement only switches once.
+  We have decided to take this restrictions as especially for ShortestPath
+  queries the results are not well-defined. If you have a use-case where
```
Comment (jsteemann) [inline CHANGELOG:25]:
```suggestion
  We have decided to take these restrictions as especially for ShortestPath
  queries the results are not well-defined. If you have a use case where
```

File: arangod/Aql/Functions.cpp [hunk 3]
```diff
@@ -9211,6 +9215,41 @@ AqlValue Functions::MakeDistributeGraphInput(
...
 }
 
+#ifdef USE_ENTERPRISE
+AqlValue Functions::SelectSmartDistributeGraphInput(
+    arangodb::aql::ExpressionContext* expressionContext, AstNode const&,
+    VPackFunctionParameters const& parameters) {
+  AqlValue const& from = extractFunctionParameterValue(parameters, 0);
+  VPackSlice input = from.slice();  // will throw when wrong type
+  if (ADB_UNLIKELY(!input.isObject() ||
+                   !input.hasKey(StaticStrings::IdString) ||
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp:9225]: This line is not necessary. The following `input.get(...).isString()` will always be false if the attribute is not present. In fact, `slice.get(attribute)` will return a Slice of type None if the attribute does not exist. 
```suggestion
```

File: arangod/Aql/GraphNode.cpp [hunk 4]
```diff
@@ -202,20 +253,27 @@ GraphNode::GraphNode(ExecutionPlan* plan, ExecutionNodeId id,
         THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_ARANGO_COLLECTION_TYPE_INVALID,
                                        msg);
       }
+      if (_isDisjoint) {
+        // TODO: Alternative to "THROW" we could run a community based Query
+        // here, instead of a Disjoint one.
+        auto res = disjointTest.isCollectionAllowed(collection, dir);
+        if (res.fail()) {
+          THROW_ARANGO_EXCEPTION_MESSAGE(res.errorNumber(), res.errorMessage());
```
Comment (jsteemann) [inline arangod/Aql/GraphNode.cpp:261]:
```suggestion
          THROW_ARANGO_EXCEPTION(res);
```

File: arangod/Aql/GraphNode.cpp [hunk 5]
```diff
@@ -202,20 +253,27 @@ GraphNode::GraphNode(ExecutionPlan* plan, ExecutionNodeId id,
         THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_ARANGO_COLLECTION_TYPE_INVALID,
```
Comment (jsteemann) [inline arangod/Aql/GraphNode.cpp:253]: The error message could be misleading if the collection cannot be found. Then we should rather return TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND and an appropriate message.
But I don't insist on making this change.

File: arangod/Aql/GraphNode.cpp [hunk 6]
```diff
@@ -245,22 +303,30 @@ GraphNode::GraphNode(ExecutionPlan* plan, ExecutionNodeId id,
...
-      auto& collections = plan->getAst()->query().collections();
       if (ServerState::instance()->isRunningInCluster()) {
         auto c = ci.getCollection(_vocbase->name(), n);
+        if (_isDisjoint) {
+          // TODO: Alternative to "THROW" we could run a community based Query
+          // here, instead of a Disjoint one.
+          auto res = disjointTest.isCollectionAllowed(c, _defaultDirection);
+          if (res.fail()) {
+            THROW_ARANGO_EXCEPTION_MESSAGE(res.errorNumber(),
+                                           res.errorMessage());
```
Comment (jsteemann) [inline arangod/Aql/GraphNode.cpp:327]:
```suggestion
            THROW_ARANGO_EXCEPTION(res);
```

**Review by jsteemann — APPROVED** [review]
LGTM

File: arangod/Aql/Functions.cpp (hunk 3 above)
Comment (mchacki) [reply arangod/Aql/Functions.cpp:9225]: good point

File: arangod/Aql/GraphNode.cpp [hunk 7]
```diff
@@ -245,22 +303,30 @@ GraphNode::GraphNode(ExecutionPlan* plan, ExecutionNodeId id,
...
       if (_options->shouldExcludeEdgeCollection(n)) {
         // excluded edge collection
         continue;
       }
 
-      auto& collections = plan->getAst()->query().collections();
       if (ServerState::instance()->isRunningInCluster()) {
         auto c = ci.getCollection(_vocbase->name(), n);
+        if (_isDisjoint) {
+          // TODO: Alternative to "THROW" we could run a community based Query
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Aql/GraphNode.cpp:322]: I would like to keep this comment, this how we can get away with allowing this case, if we ever have to.

**Review by jsteemann — APPROVED** [review]

**Pull request merged**